cfg_mgmt/azure.py

```diff
@@ -0,0 +1,183 @@
+    service_principal: model.azure.AzureServicePrincipal,
+):
+
+    app = msal.ConfidentialClientApplication(
+        client_id=service_principal.client_id(),
+        authority=f'https://login.microsoftonline.com/{service_principal.tenant_id()}',
+        client_credential=service_principal.client_secret(),
+    )
+    scopes = ['https://graph.microsoft.com/.default',]
+    response = app.acquire_token_for_client(scopes=scopes)
+
+    if 'access_token' in response:
+        access_token = response['access_token']
+    elif 'error' in response:
+        raise RuntimeError(
+            f'Error authenticating with Microsoft Graph: {response.get("error_description")}'
+        )
+    else:
+        raise RuntimeError(f'Unexpected response: {response}')
+
+    return access_token
+
+
+def _to_datetime(
+    azureDateString:str,
+):
+    # Azure ISO date format has to be converted manually to datetime
+    pos = azureDateString.find('.')
+    if pos > 0:
+        azureDateString = azureDateString[:pos] + 'Z'
+
+    return datetime.datetime.strptime(azureDateString, '%Y-%m-%dT%H:%M:%SZ')
+
+
+def _get_passwords(
+    service_principal: model.azure.AzureServicePrincipal,
+    access_token: str,
+) -> typing.Iterable[PasswordCredential]:
+    response = requests.get(
+        url=f'{AZURE_APP_ROOT_URL}/{service_principal.object_id()}',
+        headers={'Authorization': f'Bearer {access_token}'},
+    )
+    response_dict = response.json()
+    if not 'passwordCredentials' in response_dict:
+        # no keys
+        return []
+
+    return [PasswordCredential(**e) for e in response_dict['passwordCredentials']]
+
+
+def _add_password(
+    service_principal: model.azure.AzureServicePrincipal,
+    display_name: str,
+    access_token: str,
+) -> PasswordCredential:
+    '''Add a new password credential to the given service principal and return the corresponding
+    PasswordCredential object
+
+    Note: This will be the only time that the generated `secretText` (i.e. the actual password) is
+    available - no other calls to the API populate this field.
+    '''
+    body = {"passwordCredential": {"displayName": display_name}}
+
+    response = requests.post(
+        f'{AZURE_APP_ROOT_URL}/{service_principal.object_id()}/addPassword',
+        json=body,
+        headers={'Authorization': f'Bearer {access_token}'},
+    )
+    response.raise_for_status()
+    response_dict = response.json()
+
+    # filter metadata fields from the returned response (by filtering unexpected fields).
+    dataclass_fields = PasswordCredential.__annotations__.keys()
+    filtered_response = {k:v for k,v in response_dict.items() if k in dataclass_fields}
+
+    return PasswordCredential(**filtered_response)
+
+
+def _remove_password(
+    service_principal: model.azure.AzureServicePrincipal,
+    key_id: str,
+    access_token: str,
+):
+    response = requests.post(
+        f'{AZURE_APP_ROOT_URL}/{service_principal.object_id()}/removePassword',
```

Comment:
considering the amount of routes, I would suggest to use a routes-class, as we do in other api-clients - wdyt?

as a general remark, please use dependency-injection. base-url should be a function parameter. initialise it w/ a default value for convenience, fine w/ me. but avoid consuming from global/module scope, please.